### ISSUE TITLE
Prepare for addition of engine-specific router service

### ISSUE THREAD

**Issue by dgeb** [ember-engines/ember-engines#756]
Engines pre-date the introduction of the `router` service. Ideally, when that service was introduced, engines should have simultaneously introduced their own engine-specific `router` service to provide restricted access to an engine's own routes. Unfortunately, this was not done, and engines instead unintentionally inherited their host's `router` service, breaking the encapsulation that is a goal of this project.  

@villander has made a lot of progress working on an engine-specific router service in #669. However, because many apps currently make use of the host's `router` service, we can't simply swap in a new `router` service with an entirely different scope without breaking those apps.

Before introducing the new `router` service, we need to deprecate the current `router` service in a semver-compliant way, probably as an 0.8.x release. This will clear the way for complete removal of the `router` service in 0.9.x, followed by a re-introduction of an engine-specific `router` service in a release >= 0.10. 

The initial deprecation work involves the following tasks:

* Register a new `router` service that proxies all its methods to the host's `router` service. Every method invocation that's passed should include a deprecation warning.
* Create a deprecation documentation page about the router service on ember-engines.com. Provide suggestions for how to register the host's router as `hostRouter` and/or the root router as `appRouter`. However, we don't want to automatically create either of these services in an engine - the decision should be the developer's.

We should also consider allowing an experimental build flag to allow usage of an alpha engine-specific router which we can iterate upon starting in 0.9. This approach should allow opt-in usage of a router service similar to that in #669 for developers who are willing to experiment and endure some breaking changes.

This approach was discussed in February's engines team meeting with @rwjblue and @buschtoens. Please let me know if this ticket needs any adjustments / further discussion before we begin the work. @villander has already discussed reusing much of the work in #669 to deprecate the current `router` service.

**Comment by buschtoens** [comment]
Sounds exactly like what I recall from the meeting. 👍🏼 

I have a few minor notes though:

> - Register a new `router` service that proxies all its methods to the host's `router` service. Every method invocation that's passed should include a deprecation warning.

Unconditionally creating a proxy `router` service in every engine may cause unintended side-effects, e.g. some addon or user code may check for the existence of a `router` service and use different adapters then. Considering that addons that deal with routing need to special-handle engines in one way or another, if they want to be compatible, the likelihood is not zero, that the mere presence of a `router` service could change the behavior.

On the other hand, this may be super-edge-case-y, SemVer-lawyering-wise still okay and not even a real issue.

We could however only setup the proxy `router` service, when a service with that name is actually injected into the engine.

Which is a nice segue into my next nitpick:

> [...] proxies all its methods to the _host's_ `router` service [...]

This would not work for these two scenarios:

- Engine implements and registers the `router` service by itself. It is not an external injection.
- The host [remaps / aliases](https://ember-engines.com/docs/services) the `router` service name and e.g. injects `mock-router` as `router`.
  ```ts
  engines = {
    superBlog: {
      dependencies: {
        services: [
          // name in superBlog engine: 'store'
          // name in host: 'store'
          'store',

          // name in superBlog engine: 'router'
          // name in host: 'mock-router'
          { 'router': 'mock-router' }
        ]
      }
    }
  };
  ```

Instead of always "blindly" proxying to the host's `router` service, I would prefer, if we could somehow "wrap" the actual `router` service instance, when it gets registered / injected. This way we would:
- only make `router` available in engines, where it already was available
- don't accidentally break custom interfaces introduced by users, who use some sort of custom `router` service, i.e. not the `app-router`.

For instance, at @ClarkSource we've been using our own make-shift engine-specific router service as `router` inside many engines for 2½ years now, basically ever since I opened #587.

--- 

> We should also consider allowing an experimental build flag to allow usage of an alpha engine-specific router which we can iterate upon starting in 0.9. This approach should allow opt-in usage of a router service similar to that in #669 for developers who are willing to experiment and endure some breaking changes.

Big 👍🏼.

**Comment by rwjblue** [comment]
Ya, sounds like a good summary here to me.

**Comment by villander** [comment]
Great summary, thanks @dgeb

**Comment by dgeb** [comment]
Closing now, since this has effectively been resolved via the deprecation introduced in https://github.com/ember-engines/ember-engines/pull/764. As discussed in that PR, the assumption about automatic injection of the host's `router` was incorrect. This is greatly simplifying, since we only needed to deprecate injection of the `router` service and no longer need to deal with the complexities of a proxy router (which @buschtoens mentions above).
